nucliadb/nucliadb/train/tests/utils.py

```diff
@@ -0,0 +1,117 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+from typing import AsyncIterator, Union, overload
+
+import aiohttp
+from nucliadb_protos.dataset_pb2 import (
+    FieldClassificationBatch,
+    ImageClassificationBatch,
+    ParagraphClassificationBatch,
+    ParagraphStreamingBatch,
+    SentenceClassificationBatch,
+    TokenClassificationBatch,
+)
+
+TrainBatchType = Union[
+    type[FieldClassificationBatch],
+    type[ImageClassificationBatch],
+    type[ParagraphClassificationBatch],
+    type[ParagraphStreamingBatch],
+    type[SentenceClassificationBatch],
+    type[TokenClassificationBatch],
+]
+
+TrainBatch = Union[
+    FieldClassificationBatch,
+    ImageClassificationBatch,
+    ParagraphClassificationBatch,
+    ParagraphStreamingBatch,
+    SentenceClassificationBatch,
+    TokenClassificationBatch,
+]
+
+# NOTE: we use def instead of async def to make mypy happy. Otherwise, it
+# considers the overloaded functions as corountines returning async iterators
+# instead of async iterators themselves and complains about it
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[FieldClassificationBatch],
+) -> AsyncIterator[FieldClassificationBatch]:
+    ...
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[ImageClassificationBatch],
+) -> AsyncIterator[ImageClassificationBatch]:
+    ...
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[ParagraphClassificationBatch],
+) -> AsyncIterator[ParagraphClassificationBatch]:
+    ...
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[ParagraphStreamingBatch],
+) -> AsyncIterator[ParagraphStreamingBatch]:
+    ...
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[SentenceClassificationBatch],
+) -> AsyncIterator[SentenceClassificationBatch]:
+    ...
+
+
+@overload
+def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: type[TokenClassificationBatch],
+) -> AsyncIterator[TokenClassificationBatch]:
+    ...
+
+
+async def get_batches_from_train_response_stream(
+    response: aiohttp.ClientResponse,
+    pb_klass: TrainBatchType,
+) -> AsyncIterator[TrainBatch]:
```

Comment:
This in fact should be a `AsyncGenerator[TrainBatch, None]`, and I think the overloads are not needed.

Reply:
I can change it to `AsyncGenerator` but we still need the overloads to avoid mypy complaining about field access that are not in all protobufs. For example, the `ImageClassificationBatch` has a field `selections` that any other have.